core/schemas/model.py

```diff
@@ -120,46 +125,148 @@ class YetiModel(YetiBaseModel):
     aggregated_links: dict[str, dict[str, int]] | None = None
 
 
-class YetiTagModel(YetiModel):
-    _tags: dict[str, TagRelationship] = {}
+class YetiTagInstance(YetiModel):
+    name: str
+    last_seen: datetime.datetime
+    expires: datetime.datetime | None = None
+    fresh: bool
 
-    def __init__(self, **data):
-        super().__init__(**data)
-        for tag_name, value in data.get("tags", {}).items():
-            self._tags[tag_name] = TagRelationship(**value)
 
-    @computed_field(return_type=dict[str, TagRelationship])
-    @property
-    def tags(self):
-        return self._tags
+def normalize_name(tag_name: str) -> str:
+    nfkd_form = unicodedata.normalize("NFKD", tag_name)
+    nfkd_form.encode("ASCII", "ignore").decode("UTF-8")
+    tag_name = "".join([c for c in nfkd_form if not unicodedata.combining(c)])
+    tag_name = tag_name.strip().lower()
+    tag_name = re.sub(r"\s+", "_", tag_name)
+    tag_name = re.sub(r"[^a-zA-Z0-9_.:-]", "", tag_name)
+    return tag_name
 
-    def get_tags(self) -> list[tuple[TagRelationship, "Tag"]]:
-        """Returns the tags linked to this object.
 
-        Returns:
-          A list of tuples (TagRelationship, Tag) representing each tag linked
-          to this object.
-        """
-        from core.schemas.graph import TagRelationship
-        from core.schemas.tag import Tag
+class YetiTagModel(YetiModel):
+    tags: list[YetiTagInstance] = []
+
+    def tag(
+        self,
+        tags: list[str],
+        strict: bool = False,
+        normalized: bool = True,
+        expiration: datetime.timedelta | None = None,
+    ):
+        """Adds or updates a tag in an object."""
+        from core.schemas import tag
+
+        if not isinstance(tags, (list, set, tuple)):
+            raise ValueError("Tags must be of type list, set or tuple.")
+
+        old_tags = [tag.name for tag in self.tags]
+        actual_tags = []
```

Comment:
use a set rather than a list to remove duplicates?

Reply:
I thought of doing this, but there should be no duplicates in `self.tags`, ever. If there are, then I think we'd rather have it show than secretly deduplicate them?

Reply:
Let's keep it this way. I was thinking of API calls where several identical tags could be added.